Makefile

```diff
@@ -35,15 +36,56 @@ OS?=linux
 ARCH?=amd64
 OSVERSION?=amazon
 
-IMAGE_PLATFORMS?=linux/arm64,linux/amd64
+ALL_OS?=linux
+ALL_ARCH_linux?=amd64 arm64
+ALL_OSVERSION_linux?=amazon
+ALL_OS_ARCH_OSVERSION_linux=$(foreach arch, $(ALL_ARCH_linux), $(foreach osversion, ${ALL_OSVERSION_linux}, linux-$(arch)-${osversion}))
+
+ALL_OS_ARCH_OSVERSION=$(foreach os, $(ALL_OS), ${ALL_OS_ARCH_OSVERSION_${os}})
+
+PLATFORM?=linux/amd64,linux/arm64
+
+# split words on hyphen, access by 1-index
+word-hyphen = $(word $2,$(subst -, ,$1))
 
 .EXPORT_ALL_VARIABLES:
 
-.PHONY: aws-fsx-csi-driver
-aws-fsx-csi-driver:
-	mkdir -p bin
-	@echo GOARCH:${GOARCH}
-	CGO_ENABLED=0 GOOS=linux go build -ldflags ${LDFLAGS} -o bin/aws-fsx-csi-driver ./cmd/
+.PHONY: linux/$(ARCH) bin/aws-fsx-csi-driver
+linux/$(ARCH): bin/aws-fsx-csi-driver
+bin/aws-fsx-csi-driver: | bin
+	CGO_ENABLED=0 GOOS=linux GOARCH=$(ARCH) go build -ldflags ${LDFLAGS} -o bin/aws-fsx-csi-driver ./cmd/
+
+all: all-image-docker
+
+all-push:
+	docker buildx build \
+		--platform=$(PLATFORM) \
+		--progress=plain \
+		--target=$(OS)-$(OSVERSION) \
+		--output=type=registry \
+		-t=$(IMAGE):$(TAG) \
+		.
+	touch $@
+
+all-image-docker: $(addprefix sub-image-docker-,$(ALL_OS_ARCH_OSVERSION_linux))
+
+sub-image-%:
+	$(MAKE) OUTPUT_TYPE=$(call word-hyphen,$*,1) OS=$(call word-hyphen,$*,2) ARCH=$(call word-hyphen,$*,3) OSVERSION=$(call word-hyphen,$*,4) image
+
+image: .image-$(TAG)-$(OS)-$(ARCH)-$(OSVERSION)
```

Comment:
the main userfacing ones should be make all and make all-push (which is executed by github workflow). But make image has a usecase during CI when we want to only build linux/amd64

The vars are always defaulted by ?=

And i know i can mark them phony but it's annoying to do it for every rule. in particular, the .image should not be phony because it touches a file at the end so that the same image for the same tag+os+arch is not built again

Reply:
PHONY'd the rules I could in 2 nd commit xd